src/network/netdev/wireguard.c

```diff
@@ -0,0 +1,714 @@
+                  }
+                  r = sd_netlink_message_close_container(message);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not add wireguard allowed ip: %m");
+
+                  r = sd_netlink_message_close_container(message);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not add wireguard peer: %m");
+          }
+
+          peer_start = peer;
+          if (peer_start && !mask_start) {
+                  r = sd_netlink_message_cancel_array(message);
+                  if (r < 0)
+                          return log_netdev_error_errno(netdev, r, "Could not cancel wireguard peers: %m");
+          }
+
+          r = sd_netlink_message_close_container(message);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could close wireguard container: %m");
+
+          r = sd_netlink_send(netdev->manager->genl, message, &serial);
+          if (r < 0)
+                  return log_netdev_error_errno(netdev, r, "Could not set wireguard device: %m");
+
+        } while (peer || mask_start);
+
+        return 0;
+}
+
+static WireguardEndpoint* endpoint_free(WireguardEndpoint *e) {
+        assert(e);
+        netdev_unref(e->netdev);
+        free(e->host);
+        free(e->port);
+        free(e);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(WireguardEndpoint*, endpoint_free);
+
+static int on_resolve_retry(sd_event_source *s, usec_t usec, void *userdata) {
+        NetDev *netdev = userdata;
+        Wireguard *w;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        w->resolve_retry_event_source = sd_event_source_unref(w->resolve_retry_event_source);
+
+        w->unresolved_endpoints = w->failed_endpoints;
+        w->failed_endpoints = NULL;
+
+        resolve_endpoints(netdev);
+
+        return 0;
+}
+
+static int wireguard_resolve_handler(sd_resolve_query *q,
+                                     int ret,
+                                     const struct addrinfo *ai,
+                                     void *userdata) {
+        NetDev *netdev;
+        Wireguard *w;
+        _cleanup_(endpoint_freep) WireguardEndpoint *e;
+        int r;
+
+        assert(userdata);
+        e = userdata;
+        netdev = e->netdev;
+
+        assert(netdev);
+        w = WIREGUARD(netdev);
+        assert(w);
+
+        LIST_REMOVE(endpoints, w->unresolved_endpoints, e);
+
+        if (ret != 0) {
+                log_netdev_error(netdev, "Failed to resolve host '%s:%s': %s", e->host, e->port, gai_strerror(ret));
+                LIST_PREPEND(endpoints, w->failed_endpoints, e);
+                e = NULL;
+        } else if ((ai->ai_family == AF_INET && ai->ai_addrlen == sizeof(struct sockaddr_in)) ||
+                        (ai->ai_family == AF_INET6 && ai->ai_addrlen == sizeof(struct sockaddr_in6)))
```

Comment:
consider comparing with FAMILY_ADDRESS_SIZE() here

Reply:
Can you elaborate on that? I compare `struct sockaddr_in` that has a different size then the return value of `FAMILY_ADDRESS_SIZE`.

Reply:
`FAMILY_ADDRESS_SIZE()` is for determining the length of the enclosed IP address, which is sometimes useful, but what we want here is actually to determine the length of the entire sockaddr structure, since it needs to be either exactly `sizeof(sockaddr_in)` or `sizeof(sockaddr_in6)`. So I think this code is correct, as is.